File: ruby/lib/redis/commands/save.rb
require_relative './base'

module Redis
  module Commands
    class Save < Base
      SNAPSHOT_PATH = 'snapshot/dump.rdb'.freeze
      
      def self.command
        'save'
      end
      
      def run
        @save_pid = Process.fork do
          File.open(SNAPSHOT_PATH, 'w') do |f|
            @store.each do |key, value|
              f.write("#{key}:#{value}\n")
            end
          end
        end
        
        Redis::Resp::Serializer.serialize(:OK)
      end
    end
  end
end

File: python/src/commands/save.py
from .base import BaseCommand
from src.resp.serializer import RespSerializer
import os

class Save(BaseCommand):
  SAVE_PATH = 'snapshot/save.rdb'

  @classmethod
  def command(cls) -> bytes:
    return b'save'
  
  def run(self) -> bytes:
    self.saved_pid = os.fork()
    if self.saved_pid == 0:
      with open(self.SAVE_PATH, 'w') as f:
        for key, value in self.store.items():
          f.write(f'{key}:{value}\n')
      os._exit(0)

      # return RespSerializer.serialize(b'OK')
